src/raw/mod.rs

```diff
@@ -2781,31 +2781,34 @@ impl<T: Clone, A: Allocator + Clone> Clone for RawTable<T, A> {
                 // This leak is unavoidable: we can't try dropping more elements
                 // since this could lead to another panic and abort the process.
                 //
-                // SAFETY: We clear our table right after dropping the elements,
-                // so there is no double drop, since `items` will be equal to zero.
-                guard.drop_elements();
-
-                // Okay, we've successfully dropped all elements, so we'll just set
-                // `items` to zero (so that the `Drop` of `RawTable` doesn't try to
-                // drop all elements twice) and just forget about the guard.
-                guard.table.items = 0;
-                mem::forget(guard);
+                // SAFETY: If something gets wrong we clear our table right after
+                // dropping the elements, so there is no double drop, since `items`
+                // will be equal to zero.
+                self_.drop_elements();
 
                 // If necessary, resize our table to match the source.
-                if self.buckets() != source.buckets() {
+                if self_.buckets() != source.buckets() {
                     // Skip our drop by using ptr::write.
-                    if !self.table.is_empty_singleton() {
+                    if !self_.table.is_empty_singleton() {
                         // SAFETY: We have verified that the table is allocated.
-                        self.free_buckets();
+                        self_.free_buckets();
                     }
-                    (self as *mut Self).write(
+                    // Let's read `alloc` for reusing in new table allocator
+                    // SAFETY:
+                    // * `&mut self_.table.alloc` is valid for reading, properly
+                    //    aligned, and points to a properly initialized value as
+                    //    it is derived from a reference.
+                    //
+                    // *  We want to overwrite our own table.
+                    let alloc = ptr::read(&self_.table.alloc);
+                    (&mut **self_ as *mut Self).write(
                         // Avoid `Result::unwrap_or_else` because it bloats LLVM IR.
                         //
                         // SAFETY: This is safe as we are taking the size of an already allocated table
                         // and therefore сapacity overflow cannot occur, `self.table.buckets()` is power
                         // of two and all allocator errors will be caught inside `RawTableInner::new_uninitialized`.
                         match Self::new_uninitialized(
-                            self.table.alloc.clone(),
+                            alloc,
```

Comment:
Maybe it's better to use `alloc.clone()` here?

Reply:
No, your current code looks fine and is arguably more correct than cloning the allocator.

In the long term, I would like to remove the `Clone` bound on the allocator. It is only required due to an implementation quirk where we have two `RawTable` with an allocator each when resizing a table.

Reply:
Interesting idea.  I got an idea how to do it.  We can simply split out the part without the allocator from RawTableInner, and shove all the code that does not depend on the allocator there.  This will allow us to have two RawTableInner_Inner and one allocator at the same time (well, in theory, I haven't thought deeply yet).

Can I try to implement this?